### PR TITLE
[TableGen] Add PrintError family overload that take a print function

### PR DESCRIPTION
**Pull request by jurahul**
Add PrintError and family overload that accepts a print function. This avoids constructing potentially long strings for passing into these print functions.

### CODE REVIEW

File: llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp [hunk 1]
```diff
@@ -1600,18 +1601,13 @@ static TreePatternNode &getOperandNum(unsigned OpNo, TreePatternNode &N,
...
-  if (OpNo >= N.getNumChildren()) {
-    std::string S;
-    raw_string_ostream OS(S);
-    OS << "Invalid operand number in type constraint " << (OpNo + NumResults)
-       << " ";
-    N.print(OS);
-    PrintFatalError(S);
+  if (OpNo - NumResults >= N.getNumChildren()) {
+    PrintFatalError([&N, OpNo](raw_ostream &OS) {
+      OS << formatv("Invalid operand number {0} in type constraint ", OpNo);
```
Comment (topperc) [inline llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp:1606]: Doesn't formatv create a buffer to hold the full string before printing? Would `OS << "Invalid operand number " << OpNo << "in type constraint" be better?

**Comment by jurahul** [comment]
AFAIK it does not. It splits the format string into fragments and prints
them one-by-one.

**Comment by jurahul** [comment]
But your suggestion works as well.

File: llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp (hunk 1 above)
Comment (jurahul) [reply llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp:1606]: AFAIK it does not. It splits the format string into fragments and prints them one-by-one. Though in this case, may be what you suggested is better (in the sense what we are printing is relatively simple and does not warrant use of formatv). I'll change it.

Comment (jurahul) [reply llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp]: Done.

**Comment by jurahul** [comment]
I removed the use of formatv() in the new code as it was simple enough to not warrant the need of formatv().

File: llvm/lib/TableGen/Error.cpp [hunk 2]
```diff
@@ -40,12 +40,22 @@ static void PrintMessage(ArrayRef<SMLoc> Loc, SourceMgr::DiagKind Kind,
                         "instantiated from multiclass");
 }
 
+[[noreturn]] inline static void exit() {
```
Comment (tmatheson-arm) [inline llvm/lib/TableGen/Error.cpp:43]: Please could you pick a more descriptive name than `exit`, something that hints at what it does.

File: llvm/utils/TableGen/IntrinsicEmitter.cpp [hunk 3]
```diff
@@ -122,21 +121,24 @@ void IntrinsicEmitter::EmitEnumInfo(const CodeGenIntrinsicTable &Ints,
...
   if (!Set) {
-    std::vector<std::string> KnownTargets;
-    for (const auto &Target : Ints.Targets)
-      if (!Target.Name.empty())
-        KnownTargets.push_back(Target.Name.str());
-    PrintFatalError("tried to generate intrinsics for unknown target " +
-                    IntrinsicPrefix +
-                    "\nKnown targets are: " + join(KnownTargets, ", ") + "\n");
+    // The first entry is for target independent intrinsics, so drop it.
+    auto KnowTargets = ArrayRef<TargetSet>(Ints.Targets).drop_front();
```
Comment (tmatheson-arm) [inline llvm/utils/TableGen/IntrinsicEmitter.cpp:134]: Mixing in unrelated refactoring like this makes it harder to review. Same with the `OpNo -= NumResults` change above.

**Review by tommat01 — APPROVED** [review]
LGTM

File: llvm/utils/TableGen/IntrinsicEmitter.cpp (hunk 3 above)
Comment (jurahul) [reply llvm/utils/TableGen/IntrinsicEmitter.cpp:134]: This one isn't unrelated, this was the motivation to add this function. Currently it concatenates a bunch of strings to pass into PrintFatalError().

File: llvm/lib/TableGen/Error.cpp (hunk 2 above)
Comment (jurahul) [reply llvm/lib/TableGen/Error.cpp:43]: Will do

Comment (jurahul) [reply llvm/lib/TableGen/Error.cpp]: Done.

**Comment by jurahul** [comment]
I also removed the "OpNo -= NumResults" refactor.
